### ISSUE TITLE
Java Routing DSL does not work with Java 12 onwards

### ISSUE THREAD

**Issue by amanteaux** [playframework/playframework#10055]
This is the same as issue #5136: from Java 9 onward users see this error:
```
Exception in thread "main" java.lang.IllegalArgumentException: Don't know how to bind argument of type class net.jodah.typetools.TypeResolver$Unknown
```

It looks like it is related to https://github.com/jhalterman/typetools/issues/52, but since this issue is inactive since April 2019, maybe it is time to move away from typetools library. Is this library really required?

### Play Version
2.8.1

**Comment by marcospereira** [comment]
Hi @amanteaux,

Can you provide a reproducer? I tried with Java 11 and it works like expected. This is the Java version I'm using:

```
openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.6+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.6+10, mixed mode)
```

**Comment by amanteaux** [comment]
Hi @marcospereira,

Sorry, my mistake about Java 9, I was actually using Java 12...

The issue occurs with Java 12 and Java 13.
I created a minimal example to reproduce it at: https://github.com/amanteaux/play-dsl-java12
This example contains a single class that raises the error: https://github.com/amanteaux/play-dsl-java12/blob/master/src/main/java/playdsltest/Application.java

Do not hesitate if you need more details.

I updated the title of the issue.

**Comment by marcospereira** [comment]
Thanks, @amanteaux.

I was able to reproduce the problem using your project and Java 13 (the version I had installed already). The full stack trace is:

```
java.lang.IllegalArgumentException: Don't know how to bind argument of type class net.jodah.typetools.TypeResolver$Unknown
    at play.routing.RoutingDsl.pathBindableFor (RoutingDsl.java:289)
    at play.routing.RoutingDsl.with (RoutingDsl.java:238)
    at play.routing.RoutingDsl.access$000 (RoutingDsl.java:86)
    at play.routing.RoutingDsl$PathPatternMatcher.build (RoutingDsl.java:443)
    at play.routing.RoutingDsl$PathPatternMatcher.routingTo (RoutingDsl.java:364)
    at playdsltest.Application.lambda$main$1 (Application.java:18)
    at play.core.server.JavaServerHelper$$anon$3.router (Server.scala:346)
    at play.BuiltInComponentsFromContext.createHttpRequestHandler (BuiltInComponentsFromContext.java:218)
    at play.libs.F$LazySupplier.get (F.java:357)
    at play.BuiltInComponentsFromContext.httpRequestHandler (BuiltInComponentsFromContext.java:206)
    at play.BuiltInComponentsFromContext.createApplication (BuiltInComponentsFromContext.java:139)
    at play.libs.F$LazySupplier.get (F.java:357)
    at play.BuiltInComponentsFromContext.application (BuiltInComponentsFromContext.java:126)
    at play.core.server.JavaServerHelper$.forRouter (Server.scala:345)
    at play.core.server.JavaServerHelper.forRouter (Server.scala)
    at play.server.Server$Builder.build (Server.java:231)
    at play.server.Server.forRouter (Server.java:135)
    at playdsltest.Application.main (Application.java:12)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:567)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:282)
    at java.lang.Thread.run (Thread.java:830)
```

And, yes, it looks like it is related to jhalterman/typetools#52 and [reflection changes in Java 12](https://bugs.openjdk.java.net/browse/JDK-8210522).

> Is this library really required?

Yes, and I don't think it is a good idea for us to bake our own reflection tools, so this dependency will be kept for a while. I think the way to go here is to fix typetools to work with Java 12 if possible, and then update Play to a newer release.

Finally, Play does not support Java 12+, but I will add the proper labels in this issue so that we can better track things when adding support for the next LTS of the JDK.

**Comment by mkurz** [comment]
Fixed by #10814
I tested with the reproducer project and can confirm it is working now.
